core/src/types/blocking_read/blocking_reader.rs

```diff
@@ -76,21 +83,67 @@ impl BlockingReader {
 
         let mut bufs = Vec::new();
         let mut offset = start;
+        let concurrent = self.options.concurrent() as u64;
+
+        let (interval_size, mut intervals) = end
+            .map(|end| {
+                // let interval_size = (end - start + concurrent - 1) / concurrent;
+                let interval_size = (end - start) / concurrent;
+                let remainder = (end - start) % concurrent;
+                let intervals: Vec<(u64, u64)> = (0..concurrent)
+                    .map(|i| {
+                        let interval_start = start + i * interval_size + remainder.min(i);
+                        let interval_end =
+                            interval_start + interval_size + if i < remainder { 1 } else { 0 };
+                        (interval_start, interval_end)
+                    })
+                    .filter(|(interval_start, interval_end)| interval_start != interval_end)
+                    .collect();
+                (interval_size, intervals)
+            })
+            .unwrap_or({
+                // TODO: use service preferred io size instead.
+                let interval_size = 4 * 1024 * 1024;
+                let intervals: Vec<(u64, u64)> = (0..concurrent)
+                    .map(|i| {
+                        let current = start + i * interval_size;
+                        (current, current + interval_size)
+                    })
+                    .collect();
+                (interval_size, intervals)
+            });
 
         loop {
-            // TODO: use service preferred io size instead.
-            let limit = end.map(|end| end - offset).unwrap_or(4 * 1024 * 1024) as usize;
-            let bs = self.inner.read_at(offset, limit)?;
-            let n = bs.remaining();
-            bufs.push(bs);
-            if n < limit {
-                return Ok(bufs.into_iter().flatten().collect());
+            let results: Vec<Result<(usize, Buffer)>> = intervals
```

Comment:
I prefer to have similiar design to `Reader` that implement a `BufferIterator`

Reply:
Can you be more specific?

Reply:
Something like this:

https://github.com/apache/opendal/blob/42db08a2740675af5ef6742d08584e77f4950616/core/src/types/read/buffer_stream.rs#L72-L106

Reply:
`read` method need return a `Result<Buffer>`, implement a `BufferIterator` how to return?

Reply:
> Or do you mean to move this part of the logic to the `BufferIterator`, to make an encapsulation?

Yes. `RangeIterator` -> `BufferIterator` -> `read` / `StdReader` / `BytesIterator`

Reply:
I submit a new one, but `read` method will be &mut.